dia-batching-server/src/api/coingecko.rs

```diff
@@ -0,0 +1,310 @@
+use std::collections::HashMap;
+use rust_decimal::Decimal;
+use serde::de::DeserializeOwned;
+use serde::{Deserialize, Serialize};
+
+use crate::api::error::CoingeckoError;
+use crate::api::Quotation;
+use crate::args::CoingeckoConfig;
+use crate::AssetSpecifier;
+
+pub struct CoingeckoPriceApi {
+	client: CoingeckoClient,
+}
+
+impl CoingeckoPriceApi {
+	pub fn new_from_config(config: CoingeckoConfig) -> Self {
+		let api_key = config.cg_api_key.expect("Please provide a CoinGecko API key");
+
+		Self::new(config.cg_host_url, api_key)
+	}
+
+	pub fn new(host_url: String, api_key: String) -> Self {
+		let client = CoingeckoClient::new(host_url, api_key);
+
+		Self { client }
+	}
+
+	pub async fn get_prices(
+		&self,
+		assets: Vec<&AssetSpecifier>,
+	) -> Result<Vec<Quotation>, CoingeckoError> {
+		let coingecko_ids = assets
+			.clone()
+			.into_iter()
+			.filter_map(|asset| {
+				let id = Self::convert_to_coingecko_id(asset);
+				match id {
+					Some(id) => Some(id),
+					None => {
+						log::warn!("Could not find CoinGecko ID for asset {:?}", asset);
+						None
+					},
+				}
+			})
+			.collect::<Vec<_>>();
+
+		let id_to_price_map =
+			self.client.price(&coingecko_ids, false, true, false, true).await.map_err(|e| {
+				CoingeckoError(format!("Couldn't query CoinGecko prices {}", e.to_string()))
+			})?;
+
+		let quotations = id_to_price_map
+			.into_iter()
+			.filter_map(|(id, price)| {
+				let asset = assets.iter().find(|asset| {
```

Comment:
What's the need to check _again_ if the `id` in _`id_to_price_map`_ is a vaild/existing CoinGecko id?
Wasn't the _`coingecko_ids`_ variable used to generate _`id_to_price_map`_ ?

Reply:
The 'problem' is that the 'id_to_price_map' does not have information about the `AssetSpecifier`. The `AssetSpecifier` holds information about the `symbol` and `name` that we want to use in the `Quotation` that we return below. The `id_to_price_map` only has Coingecko-related info, ie. the `id` used is the coingecko identifier and the `price` in the map holds the coingecko data. We only have a one-way mapping of `AssetSpecifier` -> `CoingeckoID` so we iterate over all assets in the `id_to_price_map` doing this one-way lookup and trying to find the matching asset again.